File: taskman/server/models/project.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, List
from taskman.shared.constants import ProjectStatus

class Project(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    priority: int = Field(ge=1, le=5, default=3)
    status: ProjectStatus = ProjectStatus.ACTIVE
    tags: List[str] = Field(default_factory=list)
    parent_id: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
    completed_at: Optional[datetime] = None

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }

class ProjectCreate(BaseModel):
    name: str
    description: Optional[str] = None
    priority: int = Field(ge=1, le=5, default=3)
    tags: List[str] = Field(default_factory=list)
    parent_id: Optional[str] = None

class ProjectUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    priority: Optional[int] = Field(None, ge=1, le=5)
    status: Optional[ProjectStatus] = None
    tags: Optional[List[str]] = None
    parent_id: Optional[str] = None

File: packages/core/src/models/project.ts
import { v4 as uuidv4 } from 'uuid';

export type ProjectStatus = 'active' | 'paused' | 'completed' | 'archived';

export interface Project {
  id: string;
  name: string;
  description: string | null;
  priority: number;  // 1-5
  status: ProjectStatus;
  tags: string[];
  parent_id: string | null;
  color: string | null;  // Hex color for visual distinction (e.g., "#3B82F6")
  created_at: string;  // ISO datetime string
  updated_at: string;  // ISO datetime string
  completed_at: string | null;  // ISO datetime string
  deadline?: string | null;     // ISO date string (YYYY-MM-DD) — target completion date
}

export interface ProjectCreate {
  name: string;
  description?: string | null;
  priority?: number;
  tags?: string[];
  parent_id?: string | null;
  color?: string | null;
  deadline?: string | null;
}

export interface ProjectUpdate {
  name?: string;
  description?: string | null;
  priority?: number;
  status?: ProjectStatus;
  tags?: string[];
  parent_id?: string | null;
  color?: string | null;
  deadline?: string | null;
}

export function createProject(data: ProjectCreate): Project {
  const now = new Date().toISOString();
  return {
    id: uuidv4(),
    name: data.name,
    description: data.description ?? null,
    priority: data.priority ?? 3,
    status: 'active',
    tags: data.tags ?? [],
    parent_id: data.parent_id ?? null,
    color: data.color ?? null,
    created_at: now,
    updated_at: now,
    completed_at: null,
    deadline: data.deadline ?? null,
  };
}
